checks/pinned_dependencies.go

```diff
@@ -494,6 +526,87 @@ func validateGitHubWorkflowIsFreeOfInsecureDownloads(pathfn string, content []by
 	return true, nil
 }
 
+// The only OS that this job runs on is Windows.
+func jobAlwaysRunsOnWindows(job *gitHubActionWorkflowJob) bool {
+	var jobOses []string
+	// The 'runs-on' field either lists the OS'es directly, or it can have an expression '${{ matrix.os }}' which
+	// is where the OS'es are actually listed.
+	if len(job.RunsOn) == 1 && strings.Contains(job.RunsOn[0], "matrix.os") {
+		jobOses = job.Strategy.Matrix.Os
+	} else {
+		jobOses = job.RunsOn
+	}
+	for _, os := range jobOses {
+		if !strings.HasPrefix(strings.ToLower(os), "windows-") {
+			return false
+		}
+	}
+	return true
+}
+
+// getShellForStep returns the shell that is used to run the given step.
+func getShellForStep(step *gitHubActionWorkflowStep, job *gitHubActionWorkflowJob) (string, error) {
+	// https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#using-a-specific-shell.
+	if step.Shell != "" {
+		return step.Shell, nil
+	}
+	if job.Defaults.Run.Shell != "" {
+		return job.Defaults.Run.Shell, nil
+	}
+
+	const defaultShellNonWindows = "bash"
+	const defaultShellWindows = "pwsh"
+
+	isStepWindows, err := isStepWindows(step)
+	if err != nil {
+		return "", err
+	}
+	if isStepWindows {
+		return defaultShellWindows, nil
+	}
+
+	if jobAlwaysRunsOnWindows(job) {
+		return defaultShellWindows, nil
+	}
+
+	return defaultShellNonWindows, nil
+}
+
+// isStepWindows returns true if the step will be run on Windows.
+func isStepWindows(step *gitHubActionWorkflowStep) (bool, error) {
+	// Looking for "if: runner.os == 'Windows'" (and variants)
+	matches, err := regexp.MatchString(`(?i)runner\.os\s*==\s*['"]windows['"]`, step.If)
+	if err != nil {
+		//nolint:wrapcheck
+		return false, sce.Create(sce.ErrScorecardInternal, fmt.Sprintf("error matching runner.os regex: %v", err))
```

Comment:
Instead of repeating same code multiple times run it inside a for loop?

Reply:
Oh, yeah, of course!